File: backend-spring/src/main/java/com/cronos/service/OrderService.java
package com.cronos.service;

import com.cronos.blockchain.BlockchainLedger;
import com.cronos.dto.request.CreateOrderRequest;
import com.cronos.dto.response.PaginationMeta;
import com.cronos.entity.*;
import com.cronos.repository.*;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.util.*;

@Service
@RequiredArgsConstructor
public class OrderService {
    private final OrderRepository orderRepo;
    private final OrderItemRepository orderItemRepo;
    private final ProductRepository productRepo;
    private final BlockchainLedger ledger;

    @Transactional
    public CronosOrder createOrder(UUID userId, CreateOrderRequest req) {
        CronosOrder order = CronosOrder.builder().userId(userId).shippingAddress(req.getShippingAddress()).notes(req.getNotes()).status("pending").build();
        BigDecimal totalAmount = BigDecimal.ZERO;
        List<OrderItem> items = new ArrayList<>();

        for (var item : req.getItems()) {
            Product product = productRepo.findById(item.getProductId()).orElseThrow(() -> new RuntimeException("product not found: " + item.getProductId()));
            if (product.getStock() < item.getQuantity()) throw new RuntimeException("insufficient stock for product: " + product.getName());
            BigDecimal subtotal = product.getPrice().multiply(BigDecimal.valueOf(item.getQuantity()));
            totalAmount = totalAmount.add(subtotal);
            items.add(OrderItem.builder().productId(item.getProductId()).quantity(item.getQuantity()).price(product.getPrice()).subtotal(subtotal).build());
        }
        order.setTotalAmount(totalAmount);
        orderRepo.save(order);

        for (OrderItem item : items) {
            item.setOrderId(order.getId());
            orderItemRepo.save(item);
            Product product = productRepo.findById(item.getProductId()).get();
            product.setStock(product.getStock() - item.getQuantity());
            productRepo.save(product);
        }
        ledger.recordEvent("order_created", userId, "order", order.getId(), Map.of("total_amount", totalAmount, "items_count", items.size()));
        return orderRepo.findById(order.getId()).get();
    }

    public Map<String, Object> getMyOrders(UUID userId, String pageStr, String limitStr) {
        int page = parseIntOr(pageStr, 1); int limit = parseIntOr(limitStr, 10);
        Page<CronosOrder> ordersPage = orderRepo.findByUserIdOrderByCreatedAtDesc(userId, PageRequest.of(page - 1, limit));
        PaginationMeta meta = PaginationMeta.builder().currentPage(page).perPage(limit).total(ordersPage.getTotalElements()).totalPages(ordersPage.getTotalPages()).build();
        return Map.of("orders", ordersPage.getContent(), "meta", meta);
    }

    public Map<String, Object> getAllOrders(String pageStr, String limitStr, String status) {
        int page = parseIntOr(pageStr, 1); int limit = parseIntOr(limitStr, 10);
        Page<CronosOrder> ordersPage = orderRepo.findAllFiltered(status, PageRequest.of(page - 1, limit));
        PaginationMeta meta = PaginationMeta.builder().currentPage(page).perPage(limit).total(ordersPage.getTotalElements()).totalPages(ordersPage.getTotalPages()).build();
        return Map.of("orders", ordersPage.getContent(), "meta", meta);
    }

    public CronosOrder getOrder(UUID id) { return orderRepo.findById(id).orElseThrow(() -> new RuntimeException("order not found")); }

    @Transactional
    public void cancelOrder(UUID userId, UUID orderId) {
        CronosOrder order = orderRepo.findById(orderId).orElseThrow(() -> new RuntimeException("order not found"));
        if (!order.getUserId().equals(userId)) throw new RuntimeException("unauthorized");
        if (!"pending".equals(order.getStatus())) throw new RuntimeException("only pending orders can be cancelled");
        order.setStatus("cancelled");
        if (order.getOrderItems() != null) {
            for (OrderItem item : order.getOrderItems()) {
                Product product = productRepo.findById(item.getProductId()).get();
                product.setStock(product.getStock() + item.getQuantity());
                productRepo.save(product);
            }
        }
        orderRepo.save(order);
    }

    public List<CronosOrder> getSellerOrders(UUID sellerId) { return orderRepo.findBySellerIdDistinct(sellerId); }

    private int parseIntOr(String s, int def) { try { return Integer.parseInt(s); } catch (Exception e) { return def; } }
}

File: backend/services/order_service.go
package services
import (
	"errors"
	"math"
	"strconv"
	"cronos-backend/blockchain"
	"cronos-backend/models"
	"cronos-backend/repositories"
	"cronos-backend/utils"
	"github.com/google/uuid"
)
type OrderService struct {
	OrderRepo   *repositories.OrderRepository
	ProductRepo *repositories.ProductRepository
	Ledger      *blockchain.Ledger
}
func NewOrderService(orderRepo *repositories.OrderRepository, productRepo *repositories.ProductRepository, ledger *blockchain.Ledger) *OrderService {
	return &OrderService{OrderRepo: orderRepo, ProductRepo: productRepo, Ledger: ledger}
}
type CreateOrderRequest struct {
	ShippingAddress string `json:"shipping_address" validate:"required"`
	Notes           string `json:"notes"`
	Items           []struct {
		ProductID uuid.UUID `json:"product_id" validate:"required"`
		Quantity  int       `json:"quantity" validate:"required,gt=0"`
	} `json:"items" validate:"required,min=1"`
}
func (s *OrderService) CreateOrder(userID uuid.UUID, req CreateOrderRequest) (*models.Order, error) {
	order := &models.Order{
		UserID:          userID,
		ShippingAddress: req.ShippingAddress,
		Notes:           req.Notes,
		Status:          "pending",
	}
	var totalAmount float64
	var orderItems []models.OrderItem
	for _, item := range req.Items {
		product, err := s.ProductRepo.FindByID(item.ProductID)
		if err != nil {
			return nil, errors.New("product not found: " + item.ProductID.String())
		}
		if product.Stock < item.Quantity {
			return nil, errors.New("insufficient stock for product: " + product.Name)
		}
		subtotal := product.Price * float64(item.Quantity)
		totalAmount += subtotal
		orderItems = append(orderItems, models.OrderItem{
			ProductID: item.ProductID,
			Quantity:  item.Quantity,
			Price:     product.Price,
			Subtotal:  subtotal,
		})
	}
	order.TotalAmount = totalAmount
	if err := s.OrderRepo.Create(order); err != nil {
		return nil, errors.New("failed to create order")
	}
	for i := range orderItems {
		orderItems[i].OrderID = order.ID
		if err := s.OrderRepo.CreateOrderItem(&orderItems[i]); err != nil {
			return nil, errors.New("failed to create order item")
		}
		product, _ := s.ProductRepo.FindByID(orderItems[i].ProductID)
		product.Stock -= orderItems[i].Quantity
		s.ProductRepo.Update(product)
	}
	s.Ledger.RecordEvent("order_created", userID, "order", order.ID, map[string]interface{}{
		"total_amount": totalAmount,
		"items_count":  len(orderItems),
	})
	return s.OrderRepo.FindByID(order.ID)
}
func (s *OrderService) GetMyOrders(userID uuid.UUID, pageStr, limitStr string) ([]models.Order, *utils.PaginationMeta, error) {
	page, _ := strconv.Atoi(pageStr)
	if page < 1 {
		page = 1
	}
	limit, _ := strconv.Atoi(limitStr)
	if limit < 1 {
		limit = 10
	}
	orders, total, err := s.OrderRepo.FindByUserID(userID, page, limit)
	if err != nil {
		return nil, nil, err
	}
	meta := &utils.PaginationMeta{
		CurrentPage: page,
		PerPage:     limit,
		Total:       total,
		TotalPages:  int(math.Ceil(float64(total) / float64(limit))),
	}
	return orders, meta, nil
}
func (s *OrderService) GetAllOrders(pageStr, limitStr, status string) ([]models.Order, *utils.PaginationMeta, error) {
	page, _ := strconv.Atoi(pageStr)
	if page < 1 {
		page = 1
	}
	limit, _ := strconv.Atoi(limitStr)
	if limit < 1 {
		limit = 10
	}
	orders, total, err := s.OrderRepo.FindAll(page, limit, status)
	if err != nil {
		return nil, nil, err
	}
	meta := &utils.PaginationMeta{
		CurrentPage: page,
		PerPage:     limit,
		Total:       total,
		TotalPages:  int(math.Ceil(float64(total) / float64(limit))),
	}
	return orders, meta, nil
}
func (s *OrderService) GetOrder(id uuid.UUID) (*models.Order, error) {
	return s.OrderRepo.FindByID(id)
}
func (s *OrderService) CancelOrder(userID, orderID uuid.UUID) error {
	order, err := s.OrderRepo.FindByID(orderID)
	if err != nil {
		return errors.New("order not found")
	}
	if order.UserID != userID {
		return errors.New("unauthorized")
	}
	if order.Status != "pending" {
		return errors.New("only pending orders can be cancelled")
	}
	order.Status = "cancelled"
	for _, item := range order.OrderItems {
		product, _ := s.ProductRepo.FindByID(item.ProductID)
		product.Stock += item.Quantity
		s.ProductRepo.Update(product)
	}
	return s.OrderRepo.Update(order)
}
func (s *OrderService) GetSellerOrders(sellerID uuid.UUID) ([]models.Order, error) {
	return s.OrderRepo.FindBySellerID(sellerID)
}
